pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BalancedInstanceSelector.java

```diff
@@ -69,17 +72,18 @@ Pair<Map<String, String>, Map<String, String>> select(List<String> segments, int
         if (candidates == null) {
           continue;
         }
-        List<String> candidateInstances = new ArrayList<>(candidates.size());
-        for (SegmentInstanceCandidate candidate : candidates) {
-          candidateInstances.add(candidate.getInstance());
-        }
-        String selectedInstance = _adaptiveServerSelector.select(candidateInstances);
+        SegmentInstanceCandidate candidate = _priorityGroupInstanceSelector.select(ctx, candidates);
+        // If candidates is not null, candidates is always non-empty because segments with no enabled online servers
+        // are placed in segmentStates.getUnavailableSegments()
+        assert candidate != null;
+        _brokerMetrics.addMeteredValue(BrokerMeter.REPLICA_SEG_QUERIES, 1,
```

Comment:
can we avoid this metric call here? it will get run for every segment. (same in the else block)

Reply:
Good catch. It will cause high CPU. I will create a hashmap to count the result.